Rewrite common.c: isdec*() #2951

seekaddo opened the pull request:
I have tested the three functions and everything is working well.
Thanks

magnumripper commented:
A couple of formats fail with this (see bot checks)
```
Testing: BestCrypt [Jetico BestCrypt (.jbc) PKCS12 PBE (Whirlpool / SHA-1 to SHA-512) 32/64]... (2xOMP) FAILED (valid (before init))
Testing: 7z, 7-Zip (512K iterations) [SHA256 32/32 AES]... FAILED (valid ($7z$1$19$0$1122$8$732b59fd26896e410000000000000000$2955316379$192$183$7544a3a7ec3eb99a33d80e57907e28fb8d0e140ec8
```
They might actually need more than 9 digits, I haven't looked yet. If that's it, we need to either enhance your function or stop using it in those formats.

BTW here's how you can test your work:
```
./john -test=0
./john -test-full=0
```
They test stuff differently, both should be ran to completion.

We also have a "test suite": https://github.com/magnumripper/jtrTestSuite

magnumripper commented on src/common.c:
```diff
@@ -20,155 +20,166 @@
-		atoi64[ARCH_INDEX(*pos)] = pos - itoa64;
+    memset(atoi64, 0x7F, sizeof(atoi64));
+    for (pos = itoa64; pos <= &itoa64[63]; pos++)
+        atoi64[ARCH_INDEX(*pos)] = pos - itoa64;
```
You are introducing lots and lots of whitespace violations. Did you review your own commit at all?

seekaddo replied:
@magnumripper I will correct that. My IDEA.

magnumripper replied:
Most importantly we indent with tabs, not spaces. It doesn't matter how many spaces you'd use, it's still not correct.

solardiz commented:
(magnum was quicker than me to make some of these points, but I'll post this already-typed comment anyway)

7z and bestcrypt formats failed testing at CircleCI - please look into this. Maybe they use these functions inappropriately? If so, this PR would probably need to fix that as well.

Please do not make any whitespace changes with this PR. (It looks like there might be a tab width issue as well. I normally work on JtR source files assuming 8 char tabs.)

There are too many commits here - I'd be silly to have those completely reworked commits even mentioned in our repository. magnum, what do you say?

Although someone (Jim?) suggested documenting that these functions do not specifically test for valid inputs to `atoi()`, I think that's superfluous and confusing, especially given that in JtR we rely on `int` being at least 32-bit anyway, we limit those strings to 9 digits, and we do intend to continue calling `atoi()` after this verification at least for now. So let's drop those comment lines.

magnumripper commented:
> There are too many commits here - I'd be silly to have those completely reworked commits even mentioned in our repository. magnum, what do you say?

That alone is no problem in a PR: We can squash them when merging - even from the web gui. So just keep adding commits to this one until all is fine.

seekaddo commented:
ok I will fix the tab and the comment. The commit overflows lol, I think will have to do a new fork then. I will look into the 7z and bestcrypt format.

seekaddo commented:
Oh ok @magnumripper

solardiz commented:
The problem with bestcrypt is probably with its use of `isdecu()` followed by `strtoul()` to check for decimal values corresponding to these two constants:

```
#define kBCMode_XTS                 0xBC000004
#define kBCMode_CBC                 0xBC000002
```

I think the best fix is to drop those calls to `isdecu()` and `strtoul()` in `valid()`, and instead use `strcmp()` to check for the strings corresponding to these values in decimal. Since there are only two allowed values, that should be easy. The use of `strtoul()` in `get_salt()` may be kept as-is (we already know it's one of the two valid values, so also valid for `strtoul()` and for the assignment to `uint32_t`, by the time we reach there).

solardiz commented:
In 7z, it's a 32-bit CRC, which its `valid()` accepts as either `isdecu()` or `isdec_negok()` (IIRC, it was recently discovered that some scripts produced that value as signed, and thus possibly negative), then `get_salt()` uses `atou()` (the unsigned form). That's quite bad. I think we'll need to stop using the `isdec*()` functions for that field, and should both validate and decode it differently instead (calling `atou()` on a maybe-negative value was inappropriate, even if it just happened to work).

solardiz commented:
Dhiru added the `isdec_negok()` hack with commit 8372c905547389bb704909d5165c98cf7094af7a last month. That was an OK hack to have - better than rejecting input form those scripts - but we need to do better yet by also implementing proper decoding for both cases, which would be guaranteed to work rather than just happen to work.

We also need to add test vectors with "negative" CRC, so that if our related changes fail we'll learn right away rather than only when someone with one of those older scripts complains.

solardiz commented on src/common.c:
```diff
@@ -20,155 +20,166 @@
-	int x = atoi(q);
-	sprintf(buf, "%d", x);
-	return !strcmp(q,buf);
+    if (q && *q == '-')
```
Why did you add the `q &&` here? If someone calls one of these functions with `NULL`, we should crash, not hide the caller's error.

seekaddo replied:
@solardiz oh ok. I thought we should fail on `NULL`. I will fix that.

solardiz commented on src/common.c:
```diff
@@ -20,155 +20,166 @@
-	return !strcmp(q,buf);
+    const char *p = q;
+    do {
+        if (!q || *p < '0' || *p > '9' || p - q >= 9)
```
Ditto re: `!q ||` here.

kholia commented:
> We also need to add test vectors with "negative" CRC, so that if our related changes fail we'll learn right away rather than only when someone with one of those older scripts complains.

See https://github.com/magnumripper/JohnTheRipper/pull/2981 for this.

seekaddo commented:
After making the changes to the bestCrypt and 7z, i tested with `john -test=0` @magnumripper `john -test-full=0` not working.
now crypt is failing. 

> Testing: crypt, generic crypt(3) DES [?/64]... Segmentation fault
I am trying to trace this file.

@magnumripper The testsuite you suggested is not working. 
```
root@kali:~/Downloads/testJtr/JohnTheRipper/test# ./jtrts.pl
Can't locate jtrts_inc.pm in @INC (you may need to install the jtrts_inc module) (@INC contains: /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.26.1 /usr/local/share/perl/5.26.1 /usr/lib/x86_64-linux-gnu/perl5/5.26 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.26 /usr/share/perl/5.26 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at ./jtrts.pl line 4.
BEGIN failed--compilation aborted at ./jtrts.pl line 4.
```

@jfoug  do we have to implement a new function for checking a decimal CRC32 value? or there is something we can do to fix that. As solardiz said better to have one that work.

jfoug commented:
This is why I was not 100% 'keen' on the length <= 9 'rule' for isdec()/isdecu()  I was pretty sure there were instances wehre we needed to check up to val [0,0x80000000) and uval [0, 0xFFFFFFFF]  Testing the range of [0,0x3B9AC9FF]  (i.e. < 1 billion) works for 98% of the cases, but those other 2% can not be use that function.

magnumripper commented:
> @magnumripper The testsuite you suggested is not working.

Of course it does. Like the output suggests, you need to install some perl "libs".

seekaddo commented:
@jfoug Very good detailed idea. I will work on that. so for now I should revert all the changes solardiz suggested on the bestcrypt and 7z, try to implement this for a higher range. Thanks, Jim. will start working on it.


@magnumripper I got it. You were right.

jfoug commented:
@seekaddo the code you have is fine. It is just for the string where there exactly 10 digits.  in that case, perform a check to make sure the number is < than a specific max.   Once that is done, you have a full range isdec() function.  NOTE, once done like this, you may not be able to use ```#define isdec(a) isdecu(a)``` type logic any more. a 1 return from isdecu() no longer implies a 1 return from isdec()

solardiz commented:
We've just found some real issues in those two formats due to this attempt. Are we going to merely relax the functions that enabled this?! I think the range between 9 9's and 2 GiB or 4 GiB is rather narrow to be worth enabling it everywhere, with the integer overflow risks where the value returned by `atoi()` is then manipulated slightly (within a factor of 2). Where we actually want to support very large "hash strings" (such as when e.g. an entire encrypted archive could be encoded that way), we need to go 64-bit anyway.

But I don't feel about this strongly. So I'm fine with JimF's suggestion winning; I'll just stay out of this discussion from this point on. That's actually good news to me - I got other uses for my time, and it looks like @seekaddo might be willing to contribute to our project in other ways going forward as well, so the primary mission of gaining a contributor might be accomplished. :-)

solardiz commented:
BTW, Jim is the person who wrote the `isdec*()` functions we were trying to fix here. Now Jim says that relying on undocumented behavior of `atoi()` in a format is OK. This kind of defeats the purpose of avoiding similar inside those functions. I think we either accept that our code is sloppy and relies on documented behavior, or we don't. Occasional exceptions to any rule may exist, but they better be well-reasoned rather than just arbitrary (here we follow the rules, and here we break the rules - with no clear distinction between the two cases neither in what we break nor in our need to break).

solardiz commented:
IOW, for this PR to be consistent, if we make `isdec*()` recognize all valid 32-bit integers (signed and unsigned, respectively by the different functions), we nevertheless also need to take care of the issue found in the 7z format where it tries to use `atou()` on a negative number.

magnumripper commented:
@solardiz is that for latest 7z format being able to read non-latest 7z2john hashes? I'm not sure I disagree yet, just asking.

solardiz commented:
Yes, the needed 7z fix is "for latest 7z format being able to read non-latest 7z2john hashes". This may be a rare occasion, except that now we also have a test vector for it, so if `atou()` on a given system works differently, self-test will fail even if that functionality wouldn't actually be used in a given run.

jfoug commented:
> Now Jim says that relying on undocumented behavior of atoi() in a format is OK. T

From reference:
> Return value:   Integer value corresponding to the contents of str on success. If the converted value falls out of range of corresponding return type, the return value is undefined. If no conversion can be performed, ​0​ is returned. 

So what we have within our isdec, is trying to determine if the string represents an int value, from 0 to MAX_INT of from 0 to 0x7FFFFFFF if MAX_INT is > 0x7FFFFFFF  so I see nothing undefined here.  If we are building john for a system where MAX_INT is < 0x7FFFFFFF  (such as 0x7FFF), then the problems caused by our assumption will be meaningless, compared to the 100's to 1000's of other failure points in the source.   making isdec return 1 for p[0 .. 0x80000000) causes nothing to be undefined if sizeof(int) is 8   It simply means that only smaller data (but same size as used everywhere) will ever be tested against atoi

jfoug commented:
I certainly agree here, that using atou(-500) is very bad certainly undefined.  If that is how we are using it, then we should change that.  that is 'works' on some systems is meaningless.   it returns unsigned, and -500 is not in the defined range.

jfoug commented:
Btw, atou is 100% owned code inside of jumbo, it is this  (in misc.c)

```C
unsigned atou(const char *src) {
	unsigned val;
	sscanf(src, "%u", &val);
	return val;
}
```

Now the undefined part is sscanf("-500", "%u", &val);    I am pretty sure that using %u with a negative string is fully undefined.  I could be wrong on that, but i would REALLY be surprised.

solardiz commented:
Jim, we're in agreement on this - `atoi()` of any number from INT_MIN to INT_MAX is OK. My point is that further math is very likely to result in integer overflow - e.g., what if a format adds 1 to a number obtained like that? Indeed, we can't fully protect against such overflows at `isdec*()` level, but we can probably cover the majority of cases in Dhiru's formats, etc. by limiting to 9 9's, and dealing with the few format that need larger integers there separately.

magnumripper commented:
LOL, this is certainly not a beginner's issue. We should have realised there'd be dragons inside.

magnumripper commented:
> For the check for atoi length 10, it would be code like this
>
> ```
> int isdec(char p) {
>    // current validation code here .  returns 1 if valid, or 0 if length > 10)
>    // assert at this point, we have a solid 10 digit number string, so error checking is already done.
>    // make sure number is < 0x7fffffff  (i.e. 2147493647)
>     if (p[0] > '2') return 0;
>     if (p[1] > '1') return 0;
>     if (p[2] > '4') return 0;
>     if (p[3] > '7') return 0;
>    ///.....
>     return p[9] <= '7';
> }
> ```

That does not fly at all. It would return false for 1200000000 already.

jfoug commented:
>  That does not fly at all. It would return false for 1200000000 already. And even 1008000000.

Ok,, then a simple change

```c
    if (p[0] > '2') return 0; if p[0] < '2' return 1;
    if (p[1] > '1') return 0; if p[1] == '0' return 1;
    if (p[2] > '4') return 0; if p[2]  < '4' return 1;
    if (p[3] > '7') return 0; if p[3]  < '7' return 1;
   ///.....
    return p[9] <= '7';
```

Note, code was written off the cuff, not tested at all, but you are 100% right.  We have to continually return true or false, just like preceeding this block.  Preceeding we returned true if string < 10 digits or false if above 10. We ONLY get here on exactly 10 digits.

Then we need to work our way through each character/digit.  We either say good, or bad, or 'needs more testing'.   Only on p[9] can we say <= '7';, since we are on the 1's digit and it is either good or bad here, there is no going further.

jfoug commented:
Or better yet

```C
// NOTE, still off the cuff, untested code
int atoi(char *p) {
     char *max_num = "2147493647";
     int i;
       // all tests done.  We have pure digit string of 10 characters at this point
       // assertions  
       //    1. string is 10 characters in length, exactly.
       //    2. string is made up entirely of decimal digits.
     for (i = 0; i < 10; ++i) {
           if (p[i] > max_num[i]) return 0;
           if (p[i] < max_nump[i]) return 1;
      }
      // string exactly "2147493647" Return true.
      return 1;
}
```

magnumripper commented:
> Note, code was written off the cuff, not tested at all, but you are 100% right

Sure I thought so but we've got a young Jedi listening here so we need to play safe. This super-trivial task is definitely harder than one would imagine.

@seekaddo, hear hear. Maybe you will write crappy code in 30 years from now too, just like us. But at that point it will be distinguished crappy code.

frank-dittrich commented:
@jfoug  If we have to check whether all 10 characters are digits anyway (or would you return 1 for `p[9] < '0'`?), why not just call strcmp for the exceptional case of length 10?

jfoug commented:
Actually, the strcmp() code ___IS___ the proper code (it was what my complex ```if(>)return 0; if(<) return 1; .. continue``` code was doing.

Since we have gotten to this point in the function, we know many things about the string.   1. it is all digits.   2. it is length 10.  3. it is an ASCII string with all bytes under 0x7F.  That is pretty much all we need to use strcmp() and have it work flawlessly for us.   Note, this is the first time I have used strcmp() to do integer comparison, but trust me, it is right ;)

Yes, we had mentioned to not use any std library code.  But in this case, I am pretty sure, using the defined strcmp function is preferable to using a special case 10 character hard coded version.


@magnumripper and yes, I should think more clearly before simply puking half baked information, into this thread.

seekaddo commented:
Oh My Goodness. like minded now. Sorry, I was seriously working on it. @magnumripper you were right. I was about to complain. Oh guys, thanks for all the informative feedback. I love the support. I will go back and rewind the brains. Problem solved.

@jfoug I hope @solardiz agrees with us changing now to strcmp.
With the already refined start code using additional C library will make our lives easier. lol

solardiz commented:
Dragons are part of the game. I did tell there would be a next level despite of me eventually writing those functions. ;-)

I'm fine with the `strcmp()`, to be reached only when we already know we have exactly 10 digits. This will let us complete and close this issue for now.

As a separate step (after this issue/PR), we'll need to enhance our built-in fuzzer for it to try replacing all decimal numbers in test vectors with numbers near 2^30, 2^31, 2^32. I suspect this will trigger issues in formats that my 9 9's approach would have avoided, but that's fine - we'll deal with those issues then.

jfoug commented:
> As a separate step (after this issue/PR), we'll need to enhance our built-in fuzzer for it to try replacing all decimal numbers in test vectors with numbers near 2^30, 2^31, 2^32.

I bet you are right, here, a lot of good ground to find UB. If there is any manipulation of that integer/unsigned after it was converted from the input string, then put that string close to an edge, and let the runtime of the function push it over the top into UB land.

Now, by default, UBsan does not inform about unsigned truncation/overflow, since it is 100% defined.  You 'CAN' turn it on, but by default it is off.  I bet if on, we will have 100's of overflows, but in each case, it is expected behavior.  Hell, you can not even write a LCRNG without X bit unsigned overflowing ;)

But for the numbers just over/under 2^31, you bet. IF we are handling them signed, and they overflow, then yep, we are in UB land.   Of if there are shifts being done. We want them to be > 2^30, so that any left shift causes UB

magnumripper commented on src/common.h:
```diff
@@ -129,11 +129,12 @@ int ishexlcn(const char *q, int n);
+/*Is this a valid number and within the range of [0...999999999]*/
 int isdec_negok(const char *q);
-/* is this a valid string for atou()?  atou() func == sprintf("%x",&val) */
+/* Is this a valid number and within the range of [0...999999999]*/
```
Comments no longer reflect function

seekaddo replied:
@magnumripper oops, sorry I forgot that. And In my isdec_negok function, I forgot something too. I just remembered. Good morning everyone. 🤦

solardiz commented on src/common.c:
```diff
@@ -29,146 +29,158 @@ static int initialized = 0;
+            return 0;
+    } while (*++p);
+
+    return(strcmp(q,imin) <= 0);
```
This will wrongly allow the positive value 2147483648, which is one higher than INT_MAX. I think it's best to write a shared static function that will contain this loop and the `strcmp()` call, and will accept pointers to the string to be checked and to the string containing the max number. We'll call this function differently from all three `isdec*()`, and in particular `isdec_negok()` will call it differently based on whether the original string started with a minus sign (skip the minus and use the one higher max string) or not (just `return isdec(...)` instead, which will also use the shared function internally).

seekaddo replied:
@solardiz Thanks for the feedback. I just though about it when i was in bed, but was too late. I new the comparison at was wrong. Thanks. I like your suggestion to minimize code duplication. I will do that.

@jfoug  in `isdec_negok()` when the 10 digits number has option '-' then it should be in that range of  <= 0x80000000 (-2147483648 in 2's complement) right? which makes cases like this true `-1147483648 <= -2147483648` right?

magnumripper commented:
@seekaddo are you seeing the CircleCI and Travis build bots' verdict right above ⬆️ here? I get them over half my screen but you may have to click a red x somewhere along your commits to see the results. Anyway many more formats are failing self-test now. You need to figure out why and fix that (unless your code is correct and we're actually triggering bugs in all these formats - then maybe **we** should fix it but you'd need to point it out for us).

seekaddo commented:
@magnumripper I am going through now. Trying to find out if everything in `isdec_negok()` is as expected. it is really weird, less test was failing with the <=9 range and now more even with extension. I have to through everything.

seekaddo commented:
@magnumripper I can't seem to find this module (jtrs_inc)for the testsuite. cpan output

> Skipping jtrts_inc because I couldn't find a matching namespace.
Any suggestions?

magnumripper commented:
> @magnumripper I can't seem to find this module (jtrs_inc)for the testsuite. cpan output

That module is included in the repo. It may be that you need to run this with current directory == test, and `./jtrts.pl`. If I try to run it with CWD == src and as `../test/jtrts.pl` I get the same error.

> Because on my system john -test-full=0 doesn't work

Hmm what do you mean doesn't work?

jfoug commented:
@seekaddo 

You can simply write a trivial test bed program, and know that it works.

Copy your 3 functions to a file.  Then:

```
// note, written off the cuff.  Should be close, but might not be 100%
#include <stdio.h>
#include <string.h>
#include <unistd.h>

// 3 functions placed here (isdec, isdecu, isdec_negok.  Possibly some other 
// data will also have to follow, so that the code is 100% stand alone

void testi(const char *buf, int i) {
     if (!isdec(buf)) printf ("%d/%x fails isdec(%s)\n", i, i, buf);
     if (!isdecu(buf)) printf("%d/%x fails isdecu(%s)\n", i, i, buf);
     if (!isdec_negok(buf)) printf("%d/%x fails isdec_negok(%s)\n", i, i, buf);
}
void testu(const char *buf, unsigned u) {
   // like testi() but with printf changed
}
void testU(const char *buf, unsigned long long U) {
   // like testi() but with printf changed
}
void testI(const char *buf, long long I) {
   // like testi() but with printf changed
}
int main() {
   char buf[16];
   int i;  unsigned u;  unsigned long long U; long long I;
   // test isdec*()
   for (i = -2; i < 30; ++i) {
     sprintf(buf, "%d", i);
     testi(buf, i);
  }
  // now simply keep adding test cases, especially at edge areas, like 2^31 +- 2^17, 3^32 + 20, etc
  // you should be able to KNOW that the functions work as you want them to, without having to use
  // failures of formats as a test bed.  the failures will not check this code fully.  A test bed program can.
  // the simplest test bed, would check from -2147483648(-200) to 2147483647(+200) all strings for
  // isdec_negok() and only the values out of range should show up.  the the same full range+ test could
  // be done for the other 2.  Once you do that, you know 100% for sure, that your code 100% returns
  // true for any valid string, and returns false for any non-valid string.  You would also want to test for 
  // some other things, like empty string, 0 by itself, non digits, '-' in places other than first, longer than 10
  // byte strings, etc.  strings with 10 leading digits and other non digit stuff trailing.  9, 8, 7 digit solids
  // with non digit trailing, etc.  
  // but testing your code in an outside container, will allow you to better argue your code is solid. 
}

seekaddo commented:
@jfoug hahaha, I have the same here. I tested with some sample numbers. everything is ok with that. But I want to use the official testsuite on here. I have connected my tree to travis-ci but I don't get much there. I need it here so i can use gdb on them.

@magnumripper when i run john -test-full=0 i get unknow option.

jfoug commented:
> I can't seem to find this module (jtrs_inc)for the testsuite. cpan output

jtrts_inc.pm is not in the cpan. It is simply extra code I put into here, vs poluting the jtrts.pl (I should put more there).

You MUST be in the directory of the TS to make it work.  By convention, the jtrts is placed into  ${john_devel_root}/test  (source for john would be in ${john_devel_root}/src so test is at the sibling level).  This is the way it is setup, for the ./jtrts.pl to simply work (it needs to know where the john executable is located, since it spawns it a bunch).

There may be a few perl modules which are required by jtrts.pl but those should show up early on.  The jtrts_inc.pm is somewhat of a red-herring, since it only exists in the test suite directory. I doubt there is any reason whatsoever to try to put any of this code into cpan.  It is 100% specific to testing john.

magnumripper commented:
> @magnumripper when i run john -test-full=0 i get unknow option.

I have absolutely no idea how that would happen.
```
$ ../run/john -test=0
Will run 8 OpenMP threads
Testing: descrypt, traditional crypt(3) [DES 128/128 AVX-16]... (8xOMP) PASS
Testing: bsdicrypt, BSDI crypt(3) ("_J9..", 725 iterations) [DES 128/128 AVX-16]... (8xOMP) PASS
Testing: md5crypt, crypt(3) $1$ [MD5 128/128 AVX 4x3]... (8xOMP) PASS
\esting: bcrypt ("$2a$05", 32 iterations) [Blowfish 32/64 X2]... (8xOMP) ^CWait...
Session aborted

$ ../run/john -test-full=0
Will run 8 OpenMP threads
Testing: descrypt, traditional crypt(3) [DES 128/128 AVX-16]... (8xOMP) PASS
Testing: bsdicrypt, BSDI crypt(3) ("_J9..", 725 iterations) [DES 128/128 AVX-16]... (8xOMP) PASS
\esting: md5crypt, crypt(3) $1$ [MD5 128/128 AVX 4x3]... (8xOMP) ^CWait...
Session aborted
```

jfoug commented:
> I have absolutely no idea how that would happen.

He may be running 'john'   and not ../run/john   (i.e using installed john core, and not the john-jumbo he is building testing)

magnumripper commented:
>> @magnumripper when i run john -test-full=0 i get unknow option.
>
> I have absolutely no idea how that would happen.

Wait, maybe you have some system install of "John Proper" (non Jumbo) and actually try to run `john` as opposed to eg. `./john` or `run/.john` or whatever the relative path is. You need the latter.

seekaddo commented:
@magnumripper @jfoug Yea Yea working now. Thanks.

jfoug commented:
@seekaddo before spending time testing using dbg to test 47 formats, make SURE you have run enough data through the external test bed program, and you KNOW that the functions are behaving as they should, for full range of values.

seekaddo commented:
@jfoug ok thanks for the tips. I am doing the range now `-2147483648 --> 2147483647` now

jfoug commented:
Ok, easy problem (once I started on test bed).  I will show the results, and leave the fixing up to you ;)

```c
$ cat x.c
#include <stdio.h>
#include <string.h>

int isdec_len(const char *d, const char *mxv){

    const char *p = d;
    do {
        if (*p < '0' || *p > '9' || p - d >= 10)
             return 0;
     } while (*++p);

    return strcmp(d,mxv) <= 0;
}
int isdec(const char *q) {
    const char *imax = "2147483647";
    return isdec_len(q,imax);
}

int main() {
   unsigned u;
   int i;

   for (i = -10; i < 10; ++i) {
      char buf[11];
      sprintf(buf, "%d", i);
      if (isdec(buf) == 0)
        printf ("%d (%x) fails\n", i, (unsigned)i);
   }
   for (u = 0x7ffffffd; u < 0x8000000Fu; ++u) {
      char buf[11];
      sprintf(buf, "%u", u);
      if (isdec(buf) == 0)
        printf ("%u (%x) fails\n", u, u);
   }
   return 0;
}
```

```
$ ./a
-10 (fffffff6) fails
-9 (fffffff7) fails
-8 (fffffff8) fails
-7 (fffffff9) fails
-6 (fffffffa) fails
-5 (fffffffb) fails
-4 (fffffffc) fails
-3 (fffffffd) fails
-2 (fffffffe) fails
-1 (ffffffff) fails
3 (3) fails
4 (4) fails
5 (5) fails
6 (6) fails
7 (7) fails
8 (8) fails
9 (9) fails
2147483648 (80000000) fails
2147483649 (80000001) fails
2147483650 (80000002) fails
2147483651 (80000003) fails
2147483652 (80000004) fails
2147483653 (80000005) fails
2147483654 (80000006) fails
2147483655 (80000007) fails
2147483656 (80000008) fails
2147483657 (80000009) fails
2147483658 (8000000a) fails
2147483659 (8000000b) fails
2147483660 (8000000c) fails
2147483661 (8000000d) fails
2147483662 (8000000e) fails
```

seekaddo commented:
@jfoug why should we even test `-10; i < 10;` they are not 10 digits. I want to test the 10 digits and the edges, that above the range to see what happens.
```
2147483645/027ffffffd passes isdec_negok(2147483645)
2147483646/027ffffffe passes isdec_negok(2147483646)
2147483647/027fffffff passes isdec_negok(2147483647)
2147483648 (80000000) fails
2147483648/80000000 fails isdec_negok(2147483648)
```

jfoug commented:
Hint, the only thing needing fixed, is the return statement from isdec_len.  It needs one other check.

jfoug commented:
I made that one line change, and also bumped up the int test from -10 to 101000, and reran, and this time, I get proper output:

```
$ ./a
-10 (fffffff6) fails
-9 (fffffff7) fails
-8 (fffffff8) fails
-7 (fffffff9) fails
-6 (fffffffa) fails
-5 (fffffffb) fails
-4 (fffffffc) fails
-3 (fffffffd) fails
-2 (fffffffe) fails
-1 (ffffffff) fails
2147483648 (80000000) fails
2147483649 (80000001) fails
2147483650 (80000002) fails
2147483651 (80000003) fails
2147483652 (80000004) fails
2147483653 (80000005) fails
2147483654 (80000006) fails
2147483655 (80000007) fails
2147483656 (80000008) fails
2147483657 (80000009) fails
2147483658 (8000000a) fails
2147483659 (8000000b) fails
2147483660 (8000000c) fails
2147483661 (8000000d) fails
2147483662 (8000000e) fails
```

Now, the values which should fail (and only the ones that should fail), show up.

jfoug commented:
> @jfoug why should we even test -10; i < 10; 

It is a test bed.  hell, we could test all strings from -999999999999999999999999999999999 to 999999999999999999999999999999999 just to be on the safe side ;)  but that might take a while.  in the end, we only should have valid returned for things from 0 to 0x7fffffff (of course in decimal), but that number range. All other numbers should fail, as should anything containing any character other than a decimal digit.  

That is what a test bed program is all about.   easily throwing valid and invalid data and making sure only valid stuff sticks.

jfoug commented:
What this boils down to is one of the 3 assertions is not being met in the current code.

seekaddo commented:
@jfoug with _negok the test range should be `<-2147483648 for negt  and <=2147483647 for pos` right or my mind is mixed up now.

jfoug commented:
You are not mixed up. that range for neg_ok is correct.  

2's complement numbering can seem a bit tricky when it comes to signed, but that is how it works.

One way to 'look' at it, is the negative numbers are in pascal normal array representation  (1 to max, or in this case -1 to -max).  But the positive are in C array representation   (0 to max-1)  This is not actually quite right, but it is a simple way to wrap your head around just wtf is going on, and view 0 as +0 in 2's complement.

seekaddo commented:
@jfoug thanks, I thought i was lost. so makes sense that 
`2147483648/80000000 fails isdec_negok(2147483648)` fails because we want _negok to be < 0x8000000000 when is negative.

jfoug commented:
I will wait and see your code

```c
isdec_negok( "2147483649");   // should fail
isdec_negok("-2147483649");   // should fail

isdec_negok( "2147483648");   // should fail
isdec_negok("-2147483648");   // should pass

isdec_negok( "2147483647");   // should pass
isdec_negok("-2147483647");   // should pass
// all others working towards 0 should pass
```

seekaddo commented:
Yea I do understand you. This kind of test is ok with this code. 

> Fail for isdec_negok with value --> 2147483649
> Fail for isdec_negok with value --> -2147483649
> Fail for isdec_negok with value --> 2147483648
> pass for isdec_negok with value --> -2147483648
> pass for isdec_negok with value --> 2147483647
> pass for isdec_negok with value --> -2147483647



with this code

```
int isdec_negok(const char *q)
{

    const char *imx = "2147483648";
    if(*q !='-')
        return isdec(q);
    return isdec_len(q+1,imx);
}
```

jfoug commented:
Good reuse of isdec.

seekaddo commented:
For `isdec()` testing with a range ... <= 4294967295; Now trying test with ../run/john -test=0 to see the real problem.

jfoug commented:
> For isdecu() testing with a range ... <= 4294967295; Now trying test with ../run/john -test=0 to see the real problem.

fixed

seekaddo commented:
@jfoug thanks to solardiz kicking my brains on that. Ok where actually, do mean still in the `isdec_len` I don't think so, I was thinking is still allowing 2147483648 for positive but from the output is still doing the right thing with the range <= .. because isdec is doing the job.


I need one trick here.
How do i find this file if its test fails
`Testing: scrypt (16384, 8, 1) [Salsa20/8 128/128 AVX]... (4xOMP) FAILED (valid ($ScryptKDF.pm$16384*8*1*bjZkemVmZ3lWVi42*cmBflTPsqGIbg9ZIJRTQdbic8OCUH+904TFmNPBkuEA=))`
How do i find the scrypt function

jfoug commented:
> How do i find the scrypt function

grep ScryptKDF *.c   (will tell you where that string exists)

You do not have to test the scrypt function.  ONLY the valid() function is failing within that code, so all failures are from within valid.

AGAIN, I want to stress, that you need to use your test bed code, and find what valid data is NOT being seen as valid.  if you look at the code i posted about 15 messages above, you should see data that is being listed as NOT passing, but it is within range.   Use your test bed, and do not make it 10x harder trying to debug this within john.  The test bed testing unit is all you need.

seekaddo commented:
@jfoug Yea thanks. I found an error now. Is isdec_len wasnt testing well before stcmp. I need to make sure that we have exactly 10 digits before I do strcmp. OMG, I got this. Test bed.

I don't get what you mean with this. This doesn't have to pass for isdec because is more than INT_MAX(2147483647), and also with isdec_negok. I think this values are below  <=4294967295 and should work in isdecu.

```
2147483648 (80000000) fails
2147483649 (80000001) fails
2147483650 (80000002) fails
2147483651 (80000003) fails
2147483652 (80000004) fails
2147483653 (80000005) fails
2147483654 (80000006) fails
2147483655 (80000007) fails
2147483656 (80000008) fails
2147483657 (80000009) fails
2147483658 (8000000a) fails
2147483659 (8000000b) fails
2147483660 (8000000c) fails
2147483661 (8000000d) fails
2147483662 (8000000e) fails
```

seekaddo commented:
I have refined the isdec_len now to handle special cases.


```
int isdec_len(const char *q, const char *mxv){

    const char *p = q;
    do {
        if (*p < '0' || *p > '9' || p - q >= 10)
            return 0;
    } while (*++p);
    if((p-q) < 10){
        char buf[11];
        sprintf(buf,"%010d",54);
        return strcmp(buf,mxv);
    }
    return strcmp(q,mxv) <= 0;
}

```

seekaddo commented:
I refined the method in isdec_len to handle special digit cases for correct strcmp.
```c
int isdec_len(const char *q, const char *mxv){

    const char *p = q;
    do {
        if (*p < '0' || *p > '9' || p - q >= 10)
            return 0;
    } while (*++p);
    if((p-q) < 10){
        char buf[20];
        int w = 10 - (int)(p-q);
        sprintf(buf,"%0*d",w, 0);
        sprintf(buf+w,"%s",q);
        fprintf(stderr,"buff now: %s\n",buf);
        return strcmp(buf,mxv);
    }
    return strcmp(q,mxv) <= 0;
}
```

solardiz commented:
All you needed to do was `return p - q < 10 || strcmp(q, mxv) <= 0;`

seekaddo commented:
@solardiz I tried that, it didn't pass the `../run/john -test=0`

seekaddo commented:
@solardiz I am glad I went to through the struggle to get this green button. I have learnt a lot. Is more than a semester work. All the limitations made it fun.

solardiz commented:
@seekaddo I'm glad you're so persistent and enthusiastic about this!

Regarding the code that got the green button, I'm afraid it's just luck... bad luck, actually. With so many attempts, one of them might just happen to pass the tests. I think that code is still buggy (just why would it work right? I don't get it - we need effectively a `return 1;` within that `if` block, and your code doesn't look like it'd return 1 for all numbers up to 9 digits), but perhaps we didn't have a proper test vector to catch that.

seekaddo commented:
@solardiz You are right. I will replace that if block. It looks long and ugly. With the number of attempts -> I think is because I wasn't testing my code locally well to get a better picture of everything. I hope to improve on this whole thinking thing.

jfoug commented:
@solardiz   the code does work.  what happens is a digit number < 10 characters falls through (with the null byte check).  so it did not return a failure.  But now, we have a 3 character string, and the assertion of 10 byte string to use strcmp is not satisfied.

Thus, you have shown that simply returning   if p-q < 10 || strcmp(buf, max);

Now, we have 1 check for length < 10.  Thus that has been satified.

Also, any string > 10 characters is assured to return > 0, since even if the string starts out with 'max', if it does not end in a null termination, that byte will be larger than the null, so it will be > 0

This set of 3 functions is done.

jfoug commented:
Oh, I see his code that got the green check mark. Yes I agree, that code is not good.

The simple ```return p-q<10 || strcmp(q, mrv); ``` is the proper way to finish up that function.  I just did not want to spoon feed that answer.  I kept hinting where the problem was, I just wanted him to find a solution for it himself. 

An equally valid one would have been:

```
    if (p - q < 10)
       return 1;
    return strcmp(q, mrv);
```

Not as elegant, but probably is the same compiled code in the end after optimizer.

seekaddo commented:
@jfoug Yea thanks, guys. I am really sorry for not grasping this quick as expected, the little digits confused. I after going through everything I realised it wasn't that hard, had my head clouded. @solardiz  Sorry for all the delay and trouble, I hope to improve on that.

magnumripper commented:
You still introduce whitespace violations, even unrelated to your code. Only your actual code changes should be in the commit.

seekaddo commented:
@magnumripper  I don't have any space between those three functions. And I used the original source file and added this three functions to them. I think [this](https://github.com/magnumripper/JohnTheRipper/pull/2979/commits/e14d96b6cca403759f362004f60c5fbe4058fd16) is just because I replaced the whole lengthy if clause with a shorter one.
Can you please let me where? maybe this not what you are referring to.

magnumripper commented:
The net results of this PR can be seen in the "Files changed (2)" tab at the top of this page. It'll take you [here](https://github.com/magnumripper/JohnTheRipper/pull/2979/files). As you can see it's not pretty.

Also, you still indent your own code with spaces. I told you we do not accept that.

magnumripper commented:
> I don't have any space between those three functions

You should.

seekaddo commented:
@magnumripper I will fix that. I thought my IDEA is doing tabs now. Sorry for that bro. Should I intent with Tabs and commit again? Or you will squash it out like you said earlier.

magnumripper commented:
I can squash when merging as long as you get the final result right.

If I were you, however, I'd do it on your end because it's much easier to see the actual changes you are requesting. You can do something like `git reset aee7d98^`, then `git gui` and **only pick the changes you actually want in the commit, not the irrelevant ones**. Then `git push -f` to here.

seekaddo commented:
> I can squash when merging as long as you get the final result right.

What do you mean?

I will try this out `git reset aee7d98^,` thanks.

magnumripper commented:
>> I can squash when merging as long as you get the final result right.
>
> What do you mean?

I mean once the "Files changed" tab that I mentioned show no irrelevant changes nor any whitespace violations, I can squash & merge and it will look like a single commit. But as long as you keep all the bad and unrelated stuff in there, I won't merge it.

seekaddo commented:
@magnumripper ok, I got you.

seekaddo commented:
@magnumripper Is everything good now.

magnumripper commented:
No, did you think so? Try looking at https://github.com/magnumripper/JohnTheRipper/pull/2979/files for example.

You are supposed to change only the functions we're talking about here, nothing else in these files. Also, you still indent your stuff with spaces as opposed to tabs. You seem to lack some basic knowledge about diffs and/or git and I'm not sure just what I need to explain to you.

seekaddo commented:
@magnumripper done?

seekaddo commented:
@magnumripper  I hope everything is ok now. I have fixed the tabs problem also. If anything is not in order please let me know. Thanks.

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
-	sprintf(buf, "%d", x);
-	return !strcmp(q,buf) && *q != '-';
+	const char *imax = "2147483647";
+	return isdec_len(q, imax);
```
No need to define `imax`, you can use simply:

    return isdec_len(q, "2147483647");

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
-	int x = atoi(q);
-	sprintf(buf, "%d", x);
-	return !strcmp(q,buf);
+
```
Unneeded empty line.

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
+	if (*q != '-')
+		return isdec(q);
+
+	return isdec_len(q + 1, imx);
```
Also no need to define `imx`, and you can optionally simplify the whole function body to:

    return *q == '-' ? isdec_len(q + 1, "2147483648") : isdec(q);

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
+{
+	const char *p = q;
+	do
+	{
```
We don't normally put braces on their own lines, except for braces around function bodies. So this would be:

    do {

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
+
+	return p - q < 10 || strcmp(q, mxv) <= 0;
+}
 int isdec(const char *q)
```
Need empty line before new function.

magnumripper replied:
For the record I think @jfoug cheated with that around these functions before, so @seekaddo was mislead by it.

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
+	const char *imax = "2147483647";
+	return isdec_len(q, imax);
 }
 int isdec_negok(const char *q)
```
Need empty line before new function.

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
+
+	return isdec_len(q + 1, imx);
 }
 int isdecu(const char *q)
```
Need empty line before new function.

solardiz commented on src/common.c:
```diff
@@ -151,24 +151,34 @@ size_t hexlenl(const char *q, int *extra_chars)
-	return !strcmp(q,buf);
+	const char *umax = "4294967295";
+
+	return isdec_len(q, umax);
```
No need to define `umax`, you can use simply:

    return isdec_len(q, "4294967295");

solardiz commented on src/common.h:
```diff
@@ -129,11 +129,12 @@ int ishexlcn(const char *q, int n);
 size_t hexlenu(const char *q, int *extra_chars); /* upper cased only */
-/* is this a valid string for atoi() ONLY positive numbers are valid */
+/*Is this a valid number.
+ *ONLY positive numbers are valid. */
```
Add a single white space after the asterisks at the start of comment lines, so:

```
/* Is this a valid number.
 * ONLY positive numbers are valid. */
```

solardiz commented on src/common.h:
```diff
@@ -129,11 +129,12 @@ int ishexlcn(const char *q, int n);
+/*Is this a valid number */
 int isdec_negok(const char *q);
-/* is this a valid string for atou()?  atou() func == sprintf("%x",&val) */
+/* Is this a valid number*/
```
These comments are non-descriptive (same comment for both functions, then what's the difference between them?), and besides the (lack of) whitespace before/after asterisks is sloppy.

magnumripper commented:
Getting very close now!

Fix everything Solar said, and then we may actually be able to merge this 😄

seekaddo commented:
@solardiz @magnumripper Working on it.

magnumripper commented on src/common.c:
```diff
@@ -138,37 +142,43 @@ static MAYBE_INLINE size_t _hexlen(const char *q, unsigned char dic[0x100], int
-	return !strcmp(q,buf);
-}
+	return isdec_len(q, "4294967295");
+}
```
Missing last linefeed

seekaddo commented:
@magnumripper hahaha done.

magnumripper commented:
Aaaand... merged! Thanks.

seekaddo commented:
@magnumripper aaaaah "naaa endlich" -> Finally. Thank you guys for the patience, and all the support. You are golden. @solardiz @jfoug

solardiz commented:
Thank you for staying with us through all of this, @seekaddo!

The function descriptions in the .h file still look sloppy to me, but no more so than much of the rest of jumbo, and we got to stop somewhere.

Now we should probably open a new issue for 7z format's misuse of "%u" on a negative number. I expect it will be causing self-test failures on some platforms, so the current state is not really suitable for release. @seekaddo please feel free to open that issue and optionally work on it once open, or you may look into other issues we have. Thanks again.

jfoug commented:
@seekaddo    Good job, and great that you stayed this issue to the end.  Keep this memory about of doing things the right way, vs hacking something out that works, but can not be shown to work in all places, which is the situation you started from!  :)

If you need help in other code, just ask.  I purposely tried to point you in the right direction on this bug, without spoon feeding you the final answer.  But that will not always be the case.  If you get stuck, I can easily help get you over a hump, and let you continue.   Especially large humps are just figuring out how the core code in john loads and uses the formats themselves.  Starting with a fix to a formats valid() method (like Solar listed for the undefined usage of %u for a possible signed item), is a good way to get your feet wet and see how a format works, and to fix something that really is very wrong.

seekaddo commented:
@solardiz Thanks for the support. I have really learnt a lot more than in a Semester. I love it. 
@jfoug You are right, I need to learn more and understand the john formats. I have been looking through most of the code and how it works.Thanks for the help offer, I think solardiz want me to do more of that ->  `little hacking` lol. I will look into the 7z issue @solardiz.

I have also been going through the whole `2018 opened issues`, there is a lot of cleanup work there. Taking much of those cleanup load from you guys will help too make things faster.

solardiz commented:
> I will look into the 7z issue @solardiz.

Thanks @seekaddo. You should start by creating a GitHub issue for it, so that we don't forget.

seekaddo commented:
@solardiz ok I will do that.
